R/SIRD_plot.R

```diff
@@ -1,121 +1,158 @@
-sird_plot <- function() {
-  expr <- quote({
-    sird_equations <- function(time, variables, parameters) {
-      S <- variables[1]
-      I <- variables[2]
-      R <- variables[3]
-      D <- variables[4]
-      N <- variables[5]
-      q <- variables[6]
-      dS <- (input$muBirth * N) - (input$muDeath * S) - (input$betaSIRD * ((S * I) / (N^q)))
-      dI <- (input$betaSIRD * ((S * I) / (N^q))) - (input$gammaSIRD * I) - (input$deltaSIRD * I) - (input$muDeath * I)
-      dR <- (input$gammaSIRD * I) - (input$muDeath * R)
-      dD <- (input$deltaSIRD * I)
-      dN <- dS + dI + dR
-      list(c(dS, dI, dR, dD, dN, q))
-    }
+library(deSolve)
+library(ggplot2)
 
-    sird_values <- reactive({
-      req(input$timesteps, input$betaSIRD, input$gammaSIRD, input$deltaSIRD, input$muBirth, input$muDeath)
+equationsSIRD <- function(time, variables, parameters) {
+    S <- variables[1]
+    I <- variables[2]
+    R <- variables[3]
+    D <- variables[4]
+    N <- variables[5]
+    q <- variables[6]
+    dS <- (parameters["muB"] * N) - (parameters["muD"] * S) - (parameters["beta"] * ((S * I) / (N^q)))
+    dI <- (parameters["beta"] * ((S * I) / (N^q))) - (parameters["gamma"] * I) - (parameters["delta"] * I) - (parameters["muD"] * I)
+    dR <- (parameters["gamma"] * I) - (parameters["muD"] * R)
+    dD <- (parameters["delta"] * I)
+    dN <- dS + dI + dR
+    list(c(dS, dI, dR, dD, dN, q))
+}
+
+defaultParams <- list(
+    beta = 0.001,
+    gamma = 0.1,
+    delta = 0.05,
+    muB = 0,
+    muD = 0,
+    population = 500,
+    susceptible = 499,
+    infected = 1,
+    recovered = 0,
+    dead = 0,
+    timesteps = 50,
+    q = 0
+)
+
+# For standalone usage
+# TODO: find out if there's a better way to check if the script is being 
+# run standalone or as part of a shiny app
+solveSIRD <- function() {
+    params <- defaultParams
 
-      validate(
-        need(input$betaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$gammaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$deltaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$populationSIRD > 0, "Enter a valid value for parameters"),
-        need(input$susceptibleSIRD > 0, "Enter a valid value for parameters"),
-        need(input$infectedSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$recoveredSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$deadSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$timesteps > 0, "Timesteps must be greater than 0."),
-        need(input$populationSIRD == (input$susceptibleSIRD + input$infectedSIRD + input$recoveredSIRD), "Number of persons living in all epidemic compartments must sum to the Population Count (N)")
-      )
+    source <- if (!shiny::isRunning()) params else input
```

Comment:
`input` was not defined anywhere, so this would cause an error when the function is used. Further, there would be many unused argument warnings when the function is called as expected:

```r
solveSIRD(beta = 0.03, ...) # ... here means et cetera
```

Reply:
since we are no longer worried about the script working standalone this is does not really matter but just to explain my original thought process: I assumed 'input' would be in scope here since if it was called from a shiny app this code we all be evaluated in the parent environment ( eval.parent(expr) ) but in the case that it was not called from the shiny app it would not need to check wether 'input' was in scope since the expression would evaluate to true and reduce to 'params' (unless R does not do lazy evaluation like this).

Reply:
> Since we are no longer worried about the script working standalone

The functions will still be usable, and the file, overall, can still be sourced. It will merely define functions, and otherwise not have any side effects and not produce a value.

> I assumed 'input' would be in scope here since if it was called from a shiny app this code we all be evaluated in the parent environment ( eval.parent(expr) )

That is true! I actually misunderstood what `input` was here. I believed you were trying to refer to input arguments to this function in a mistaken way, but the semantics of your code actually intended to directly access widget values.

What the intention is with these model files is that they can be used interactively, and their default arguments help in two ways. When a user simply calls the function, they will get a quick demonstration of its functionality. When the function is called with some arguments changed, the default arguments make the call easier to read. The default arguments may also help make the Shiny application easier to read.

When these functions provided by the model files are used in Shiny, the functions will be called with arguments which will override the default argument values. The `source` object was thus a clever way of using either the values from widgets or the values from a list constant.

With default function arguments, however, the point is moot. This was clever, though. Kudos.

> but in the case that it was not called from the shiny app it would not need to check wether 'input' was in scope since the expression would evaluate to true and reduce to 'params' (unless R does not do lazy evaluation like this).

R definitely does lazily evaluate! It even has language features to make it _even lazier!_ 🛋️